### PR TITLE
Enable resampling on PeriodIndex

### PR DESCRIPTION
**Pull request by max-sixty**
This allows resampling with `PeriodIndex` objects by keeping the `group` as an index rather than coercing to a DataArray (which coerces any non-native types to objects)

I'm still getting one failure around the name of the IndexVariable still being `__resample_dim__` after resample, but wanted to socialize the approach of allowing a `name` argument to `IndexVariable` - is this reasonable?

 - [x] Closes https://github.com/pydata/xarray/issues/1270
 - [x] Tests added
 - [ ] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

### CODE REVIEW

**Comment by max-sixty** [comment]
Let me know any thoughts / alternative approaches here. 

To emphasize - the issue with coercing the labels to a `DataArray` is that non-native dtypes, such as `PeriodIndex` get converted to objects, and pandas' resampling machinery (which xarray uses) requires the standard indexes.

File: xarray/core/common.py [hunk 1]
```diff
@@ -761,12 +759,12 @@ def resample(self, indexer=None, skipna=None, closed=None, label=None,
                 'errors.'
             )
 
-        group = DataArray(dim_coord, coords=dim_coord.coords,
-                          dims=dim_coord.dims, name=RESAMPLE_DIM)
+        group = IndexVariable(
+            data=self.indexes[dim], dims=[dim], name=RESAMPLE_DIM)
```
Comment (shoyer) [inline xarray/core/common.py]: I *think* we could alternatively fix this by replace `dim_coord` in the `data` argument with `dim_coord.to_index()`, e.g., compare:
```
ipdb> DataArray(dim_coord, coords=dim_coord.coords, dims=dim_coord.dims, name=RESAMPLE_DIM).to_index()
Index([2000-01-01 00:00, 2000-01-01 06:00, 2000-01-01 12:00, 2000-01-01 18:00,
       2000-01-02 00:00, 2000-01-02 06:00, 2000-01-02 12:00, 2000-01-02 18:00,
       2000-01-03 00:00, 2000-01-03 06:00],
      dtype='object', name='time')
ipdb> DataArray(dim_coord.to_index(), coords=dim_coord.coords, dims=dim_coord.dims, name=RESAMPLE_DIM).to_index()
PeriodIndex(['2000-01-01 00:00', '2000-01-01 06:00', '2000-01-01 12:00',
             '2000-01-01 18:00', '2000-01-02 00:00', '2000-01-02 06:00',
             '2000-01-02 12:00', '2000-01-02 18:00', '2000-01-03 00:00',
             '2000-01-03 06:00'],
            dtype='period[6H]', name='time', freq='6H')
```

We go to some effort to preserve pandas.Index classes when passed into DataArray/Variable but apparently don't do it consistently everywhere. So there's also probably a lower level fix somewhere (perhaps in `xarray.core.variable.as_compatible_data`?) to ensure that this happens more consistently.

File: xarray/core/variable.py [hunk 2]
```diff
@@ -1789,12 +1789,17 @@ class IndexVariable(Variable):
     unless another name is given.
     """
 
-    def __init__(self, dims, data, attrs=None, encoding=None, fastpath=False):
+    def __init__(self, dims, data, attrs=None, encoding=None, fastpath=False, name=None):
```
Comment (shoyer) [inline xarray/core/variable.py]: I would rather not add an explicit `name` parameter to `IndexVariable`.

If anything, I would think about removing the `name` property. Long term, I'd like to rename `IndexVariable` -> `ImmutableVariable` and move the pandas adapter logic into class that we can pass into the `data` argument of `Variable`.

File: xarray/core/common.py (hunk 1 above)
Comment (max-sixty) [reply xarray/core/common.py]: Great! I'll have a look

**Comment by max-sixty** [comment]
This is so old that probably it wouldn't even be a good place to start from. Sorry I never pushed it through. I'll close.

**Pull request closed without merging**